controller/src/lib.rs

```diff
@@ -632,6 +744,253 @@ impl Controller {
+                // Validate the power state transition.
+                //
+                // For now, we enforce that modules may not go directly to high
+                // power, they have to go through low-power first.
+                //
+                // We can always set modules to low power, though, since that's
+                // valid from both off and high-power, and a no-op if it's
+                // already set.
+                let current_power_state = self.power_mode(modules).await?;
+                if matches!(mode, PowerMode::High)
+                    && current_power_state
+                        .iter()
+                        .any(|(mode, _override)| mode == &PowerMode::Off)
+                {
+                    return Err(Error::InvalidPowerStateTransition);
+                }
+
+                // We need the status bits to determine if we also need to
+                // twiddle `ResetL`.
+                let status = self.status(modules).await?;
+
+                // Check whether power is enabled / good, and / or reset
+                // asserted for any of the requested modules. We need to manage
+                // those pin states to control the power. Note that this is true
+                // regardless of whether the module has software-override of
+                // power control set. That's because we want the pins and the
+                // memory map to reflect the same state, so that toggling the
+                // software override doesn't change the power state of the
+                // module, only which hardware signals it responds to.
+                let need_power_enabled = Self::filter_modules_with(modules, &status, |st| {
+                    !st.contains(Status::POWER_GOOD | Status::ENABLED)
+                })?;
+
+                // Check for any modules which need power applied, but which do
+                // _not_ already have reset asserted. We're going to assert
+                // reset on them now, but emit a warning.
+                let need_reset_deasserted =
+                    Self::filter_modules_with(modules, &status, |st| st.contains(Status::RESET))?;
+                let need_power_but_not_reset = ModuleId {
+                    fpga_id: modules.fpga_id,
+                    ports: need_power_enabled
+                        .ports
+                        .remove(&need_reset_deasserted.ports),
+                };
+                if need_power_but_not_reset.selected_transceiver_count() > 0 {
+                    warn!(
+                        self.log,
+                        "Found modules with power disabled, but reset deasserted. \
+                        It will be asserted before enabling power.";
+                        "need_power_enabled" => ?need_power_enabled,
+                        "need_reset_deasserted" => ?need_reset_deasserted,
+                        "suspicious_modules" => ?need_power_but_not_reset,
```

Comment:
Nit: This ends up printing as `.,` in `enabling power., $` when any of those cases evaluate to true.

Reply:
Yeah, `slog` auto-inserts `,` between every "element" of the log message. I can remove it.